File: webapp/src/utils/jwt.ts
import jwt, { Algorithm } from 'jsonwebtoken';
import * as config from '../config';


interface VerifyTokenResult {
    tokenExp: boolean;
    error: any;
    decoded: any;
}


function encodeToken(username: string): string {
    const payload = {
        username: username
    }
    const options = {
        algorithm: "RS256" as Algorithm,
        expiresIn: config.TOKEN_LIFETIME
    };
    const token = jwt.sign(
        payload,
        config.JWT_PRIVATE_KEY,
        options
    );
    return token;
}

function verifyToken(token: string): VerifyTokenResult {
    const options = {
        algorithms: ["RS256"] as Algorithm[],
        maxAge: config.TOKEN_LIFETIME
    }
    try {
        const decoded = jwt.verify(token, config.JWT_PUBLIC_KEY, options);
        return { tokenExp: false, error: null, decoded: decoded };
    } catch (err) {
        return { tokenExp: true, error: err, decoded: null };
    }
}

export { encodeToken, verifyToken };

File: api/src/utils/jwt.go
package utils

import (
	"RecipeBookApi/config"
	"fmt"

	"github.com/dgrijalva/jwt-go"
)

func GenerateToken(username string) string {
	token := jwt.NewWithClaims(
		jwt.SigningMethodRS256,
		jwt.MapClaims{
			"username": username,
		},
	)
	tokenString, err := token.SignedString(config.JWT_PRIVATE_KEY)
	if err != nil {
		panic(err)
	}
	return tokenString
}

func VerifyToken(tokenString string) (*jwt.Token, error) {
	token, err := jwt.Parse(tokenString, func(token *jwt.Token) (interface{}, error) {
		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
			return nil, fmt.Errorf("Unexpected signing method: %v", token.Header["alg"])
		}
		return config.JWT_PUBLIC_KEY, nil
	})
	if err != nil {
		return nil, err
	}

	if !token.Valid {
		return nil, fmt.Errorf("Can't verify token")
	}
	return token, nil
}
